runtime/include/http_session.h

```diff
@@ -0,0 +1,369 @@
+}
+
+
+static inline void
+http_session_close(struct http_session *session)
+{
+	assert(session != NULL);
+
+	return tcp_session_close(session->socket, &session->client_address);
+}
+
+/**
+ * Writes an HTTP error code to the TCP socket associated with the session
+ * Closes without writing if TCP socket would have blocked
+ * Also cleans up the HTTP session
+ */
+static inline int
+http_session_send_err_oneshot(struct http_session *session, int status_code)
+{
+	assert(session != NULL);
+	assert(status_code >= 100 && status_code <= 599);
+
+	int rc = tcp_session_send_oneshot(session->socket, http_header_build(status_code),
+	                                  http_header_len(status_code));
+	http_session_close(session);
+	http_session_free(session);
+
+	return rc;
+}
+
+static inline int
+http_session_send_response(struct http_session *session, const char *response_content_type, void_cb on_eagain)
+{
+	assert(session != NULL);
+	assert(session->response_buffer.buffer != NULL);
+	assert(response_content_type != NULL);
+
+	int rc = 0;
+
+	struct vec_u8 *response_buffer = &session->response_buffer;
+
+	/* Send HTTP Response Header and Body */
+	rc = http_header_200_write(session->socket, response_content_type, response_buffer->length);
+	if (rc < 0) goto err;
+
+	rc = tcp_session_send(session->socket, (const char *)response_buffer->buffer, response_buffer->length,
+	                      on_eagain);
+	if (rc < 0) goto err;
+
+	http_total_increment_2xx();
+	rc = 0;
+
+done:
+	return rc;
+err:
+	debuglog("Error sending to client: %s", strerror(errno));
+	rc = -1;
+	goto done;
+}
+
+
+typedef void (*http_session_receive_request_egain_cb)(struct http_session *);
+
+/**
+ * Receive and Parse the Request for the current sandbox
+ * @return 0 if message parsing complete, -1 on error, -2 if buffers run out of space
+ */
+static inline int
+http_session_receive_request(struct http_session *session, http_session_receive_request_egain_cb on_eagain)
+{
+	assert(session != NULL);
+
+	int rc = 0;
+
+	struct vec_u8 *request_buffer = &session->request_buffer;
+	assert(request_buffer->capacity > 0);
+	assert(request_buffer->length <= request_buffer->capacity);
+
+	while (!session->http_request.message_end) {
+		/* Read from the Socket */
+
+		/* Structured to closely follow usage example at https://github.com/nodejs/http-parser */
+		http_parser                *parser   = &session->http_parser;
+		const http_parser_settings *settings = http_parser_settings_get();
+
+		/* If header parsing is complete and the header specified a content-length, resize */
+		/* If http_request.body is NULL, then the request has no body (i.e. a GET request) */
+		if (session->http_request.header_end && session->http_request.body != NULL) {
+			int header_size = (uint8_t *)session->http_request.body - session->request_buffer.buffer;
+			assert(header_size > 0);
+			int required_size = header_size + session->http_request.body_length;
+
+			assert(required_size > 0);
```

Comment:
This assertion seems to be unnecessary after the previous assertion two lines above. required_size cant be <=0 at this point.